### PR TITLE
Input masking and scaling in NBEATS (following authors' implementation)

### PR DESCRIPTION
**Pull request by prolearner**
*Issue:*
NBEATS implementation differs from [the one of the authors](https://github.com/ElementAI/N-BEATS/blob/master/models/nbeats.py). In particular it lacks proper [masking of the input](https://github.com/ElementAI/N-BEATS/blob/master/models/nbeats.py#L72), which is also described in the [paper](https://arxiv.org/pdf/1905.10437.pdf) at the end of page 26.

*Description of changes:*
Added proper input masking and optional (default to False) input scaling. Input scaling is used in the [meta-learning version of NBEATS](https://arxiv.org/pdf/2002.02887.pdf) to account for time series of different magnitudes. Here, scaling is implemented similarly to DeeAR (using MeanScaler) while in the paper the input/output of the model is scaled/descaled by the maximum value in the input.

Strangely, the authors' implementation also seems to [include part of the input as one of the additive components of the final forecasts](https://github.com/ElementAI/N-BEATS/blob/master/models/nbeats.py#L69). This pull request does not include this change since it is not described in the [original NBEATS paper](https://arxiv.org/pdf/1905.10437.pdf).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by lostella** [comment]
@prolearner looks like `black` is complaining: you need to run [this version](https://github.com/awslabs/gluon-ts/blob/master/requirements/requirements-setup.txt#L1) against `src` and `test`, and commit any re-format it applies

File: src/gluonts/model/n_beats/_network.py [hunk 1]
```diff
@@ -666,7 +694,8 @@ def __init__(self, *args, **kwargs) -> None:
 
     # noinspection PyMethodOverriding,PyPep8Naming
     def hybrid_forward(
-        self, F, past_target: Tensor, future_target: Tensor = None
+            self, F, past_target: Tensor, future_target: Tensor = None,
+            past_observed_values: Tensor = None, future_observed_values: Tensor = None,
```
Comment (lostella) [inline src/gluonts/model/n_beats/_network.py]: Maybe the optional arguments (that can be `None`) should be `Optional[Tensor]`, but I'm not sure whether mypy will like it.

Also it seems like the super classes should be able to accept `None`, since these arguments are passed on to them later on

File: src/gluonts/model/n_beats/_network.py [hunk 2]
```diff
@@ -678,15 +724,23 @@ def hybrid_forward(
             Shape: (batch_size, context_length, target_dim).
         future_target
             Not used.
+        past_observed_values
+            Tensor with past observed values.
+            Shape: (batch_size, context_length, target_dim).
```
Comment (lostella) [inline src/gluonts/model/n_beats/_network.py:729]: Same as above

File: src/gluonts/model/n_beats/_network.py [hunk 3]
```diff
@@ -666,7 +707,12 @@ def __init__(self, *args, **kwargs) -> None:
 
     # noinspection PyMethodOverriding,PyPep8Naming
     def hybrid_forward(
-        self, F, past_target: Tensor, future_target: Tensor = None
+        self,
+        F,
+        past_target: Tensor,
+        past_observed_values: Tensor,
+        future_target: Optional[Tensor] = None,
+        future_observed_values: Optional[Tensor] = None,
```
Comment (lostella) [inline src/gluonts/model/n_beats/_network.py:715]: I'm wondering if this block could just avoid having these inputs, since they aren't used. I know that `future_target` was there already, but maybe that could be removed too

File: src/gluonts/model/n_beats/_network.py [hunk 4]
```diff
@@ -508,19 +517,37 @@ def __init__(
...
             # connect last block
-            return forecast + self.net_blocks[-1](backcast)
+            forecast = forecast + self.net_blocks[-1](backcast)
+
+        forecast = F.broadcast_mul(forecast, scale)
+
+        return (
+            forecast
+            if future_observed_values is None
+            else forecast * future_observed_values
```
Comment (lostella) [inline src/gluonts/model/n_beats/_network.py:549]: Maybe an idea: just return the original (unmasked) `forecast` tensor here, and let the caller mask it where needed (e.g. the training network). Could that work?

In fact: only masking the forecast is probably not enough, and the corresponding target values which are compared in the loss should be masked too. Or otherwise, one can mask the loss values directly. (But I haven't checked in the paper if this is what the authors mean -- I hope so)

With these changes, the `future_` inputs could be removed from this block probably.

**Review by lostella — COMMENTED** [review]
Thanks @prolearner! I have some comments inline

File: src/gluonts/model/n_beats/_network.py (hunk 3 above)
Comment (prolearner) [reply src/gluonts/model/n_beats/_network.py:715]: it seems that the signature of hybrid_forward must match the one of the super class (some tests fail if this is not the case). That's why probably all four inputs are needed for the hybrid_forward of all the 3 classes.

Comment (lostella) [reply src/gluonts/model/n_beats/_network.py:715]: Ah I see, it's probably `mypy` who's complaining?

Comment (prolearner) [reply src/gluonts/model/n_beats/_network.py:715]: Can't remember, might be.

File: src/gluonts/model/n_beats/_network.py (hunk 4 above)
Comment (prolearner) [reply src/gluonts/model/n_beats/_network.py]: I think it's a good idea, I moved the masking inside the loss functions. I followed the [authors' implementation](https://github.com/ElementAI/N-BEATS/blob/04f56c4ca4c144071b94089f7195b1dd606072b0/common/torch/losses.py#L45).
Still did not manage to remove future imports (did not even try since the tests were failing before).

**Review by lostella — APPROVED** [review]
Looks good to me! Thanks for fixing this!
